workflow/util/util.go

```diff
@@ -442,21 +442,51 @@ func updateWorkflowNodeByKey(wfIf v1alpha1.WorkflowInterface, workflowName strin
 			log.Fatal(err)
 		}
 
+		nodes := wf.Status.Nodes
+		if wf.Status.IsOffloadNodeStatus() {
+			if !repo.IsEnabled() {
+				return false, fmt.Errorf(sqldb.OffloadNodeStatusDisabled)
+			}
+			var err error
+			nodes, err = repo.Get(string(wf.UID), wf.GetOffloadNodeStatusVersion())
+			if err != nil {
+				return false, fmt.Errorf("unable to retrieve offloaded nodes: %s", err)
+			}
+		}
+
 		nodeUpdated := false
-		for nodeID, node := range wf.Status.Nodes {
+		for nodeID, node := range nodes {
 			if node.IsActiveSuspendNode() {
 				if selectorMatchesNode(selector, node) {
 					node.Phase = phase
 					node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
 					if len(message) > 0 {
 						node.Message = message
 					}
-					wf.Status.Nodes[nodeID] = node
+					nodes[nodeID] = node
 					nodeUpdated = true
 				}
 			}
 		}
 		if nodeUpdated {
+			if wf.Status.IsOffloadNodeStatus() {
+				if !repo.IsEnabled() {
+					return false, fmt.Errorf(sqldb.OffloadNodeStatusDisabled)
+				}
+				offloadVersion, err := repo.Save(string(wf.UID), wf.Namespace, nodes)
+				if err != nil {
+					return false, fmt.Errorf("unable to save offloaded nodes: %s", err)
+				}
+				wf.Status.OffloadNodeStatusVersion = offloadVersion
+				wf.Status.CompressedNodes = ""
+				wf.Status.Nodes = nil
+			}
+
+			err = packer.CompressWorkflowIfNeeded(wf)
```

Comment:
This ordering is incorrect, `CompressWorkflowIfNeeded` should always come before `repo.Save`. 

This file misses unit tests. Would you please be able to add them? "No - it is too hard" would be a fine answer.

Reply:
In terms of the ordering, wouldn't that result in it always compressing the nodes if they're too big, whereas currently if offload is enabled for the workflow then it won't try to compress the nodes.

By missing tests, are you talking about adding a test to util_test.go that covers the updateWorkflowNodeByKey function?

Reply:
you should be able to see this pattern elsewhere - compression can fail - then we offload